File: src/Contract/Task.php
<?php
declare(strict_types=1);

namespace Cw\Contract;

/**
 * 通用任务载荷契约（Redis Stream 消息的 p 字段内容）。
 *
 * Runtime（Producer / Worker / RedisStore）只理解这里的结构；
 * 任何数据源特定语义（协议、字段、分页参数）都封装在 SourceAdapter 内。
 *
 * 载荷示例（与 flag.md §4 对齐）：
 * {
 *   "job_id":    "job_xxx",
 *   "source":    "shikues" | "maccms",
 *   "entity":    "model" | "vod",        // 采集对象类型
 *   "operation": "list",
 *   "cursor":    { "unit_id": "3", "unit_name": "肖特基二极管", "page": 1 },
 *   "params":    { "limit": 15, "t": 1 },  // adapter 专用，可缺省
 *   "target_pages": 3                       // 运行时策略（可缺省）
 * }
 *
 * cursor 是一个“可独立消费的游标单元”：
 *   - 一个 cursor = (source, entity, unit) 的某一次分页采集；
 *   - page 由 Worker 逐页推进，unit_id/unit_name 用于定位游标 Hash 与可读日志。
 */
final class Task
{
    /** 列表采集操作（V1 只做 list；detail/export 走 operation 扩展） */
    public const OP_LIST = 'list';

    /**
     * 由播种发现的 unit 生成首页任务。
     *
     * @param array  $unit         ['entity','unit_id','unit_name','params'?(可选)]
     * @param string $source       数据源标识
     * @param int    $page         起始页（默认 1）
     * @param int|null $targetPages 运行时抓取上限；null 表示由 Worker 读全局配置
     * @param string $jobId        所属 crawl_jobs 主键（可空）
     * @return array 载荷
     */
    public static function home(array $unit, string $source, int $page = 1, ?int $targetPages = null, ?string $jobId = null): array
    {
        $t = [
            'source'    => $source,
            'entity'    => (string)($unit['entity'] ?? ''),
            'operation' => self::OP_LIST,
            'cursor'    => [
                'unit_id'   => (string)($unit['unit_id'] ?? ''),
                'unit_name' => (string)($unit['unit_name'] ?? ''),
                'page'      => max(1, $page),
            ],
        ];
        $params = (array)($unit['params'] ?? []);
        if ($params) {
            $t['params'] = $params;
        }
        if ($targetPages !== null && $targetPages > 0) {
            $t['target_pages'] = $targetPages;
        }
        if ($jobId !== null && $jobId !== '') {
            $t['job_id'] = $jobId;
        }
        return $t;
    }

    /** 生成下一页任务（复用同 unit 的源/实体/参数，仅推进 page） */
    public static function next(array $task): array
    {
        $next = $task;
        $next['cursor']['page'] = ((int)($task['cursor']['page'] ?? 1)) + 1;
        return $next;
    }

    /** 校验载荷是否为一个合法的 list 任务 */
    public static function isValidList(array $p): bool
    {
        return isset($p['source'], $p['entity'], $p['operation'], $p['cursor'])
            && (string)$p['operation'] === self::OP_LIST
            && (string)($p['source'] ?? '') !== ''
            && (string)($p['entity'] ?? '') !== ''
            && (string)($p['cursor']['unit_id'] ?? '') !== ''
            && (int)($p['cursor']['page'] ?? 0) > 0;
    }

    /** 游标 Hash 键：cur:{source}:{entity}:{unit_id}（Reset/运维可统一 SCAN） */
    public static function cursorKey(array $p): string
    {
        return 'cur:' . $p['source'] . ':' . $p['entity'] . ':' . $p['cursor']['unit_id'];
    }

    /** 幂等“追加下一页”闸门键：addlock:{source}:{entity}:{unit_id}:{page} */
    public static function nextPageLockKey(array $p): string
    {
        return 'addlock:' . $p['source'] . ':' . $p['entity'] . ':' . $p['cursor']['unit_id'] . ':' . (int)$p['cursor']['page'];
    }

    /** 可读日志描述：source[unit_name] p{page} */
    public static function describe(array $p): string
    {
        $unit = (string)($p['cursor']['unit_name'] ?? $p['cursor']['unit_id'] ?? '');
        return sprintf('[%s|%s|p%d]', $p['source'], $unit, (int)($p['cursor']['page'] ?? 0));
    }

    /** 目标页上限：显式 target_pages 优先，否则用全局缺省 */
    public static function targetPages(array $p, int $fallback): int
    {
        $tp = (int)($p['target_pages'] ?? 0);
        return $tp > 0 ? $tp : $fallback;
    }
}

File: goKit/internal/domain/task/task.go
// Package task 定义通用任务载荷契约（Redis Stream 消息 p 字段的内容）。
//
// Runtime（Producer / Worker / RedisStore）只理解本结构；
// 任何数据源特定语义（协议、字段、分页参数）都封装在 Adapter 内。
//
// 载荷示例：
//
//	{
//	  "job_id":    "job-20260911...",
//	  "source":    "shikues",
//	  "entity":    "model",
//	  "operation": "list",
//	  "cursor":    {"unit_id": "3", "unit_name": "肖特基二极管", "page": 1},
//	  "params":    {"limit": 15},
//	  "target_pages": 3
//	}
package task

import (
	"encoding/json"
	"fmt"
)

// OpList 列表采集操作（V1 只做 list；detail/export 走 Operation 扩展）。
const OpList = "list"

// Cursor 是任务载荷内的分页游标（同时用于定位 Redis 游标 Hash）。
type Cursor struct {
	UnitID   string `json:"unit_id"`
	UnitName string `json:"unit_name"`
	Page     int    `json:"page"`
}

// Payload 是一个可独立消费的任务。
type Payload struct {
	JobID       string         `json:"job_id,omitempty"`
	Source      string         `json:"source"`
	Entity      string         `json:"entity"`
	Operation   string         `json:"operation"`
	Cursor      Cursor         `json:"cursor"`
	Params      map[string]any `json:"params,omitempty"`
	TargetPages int            `json:"target_pages,omitempty"`
}

// Unit 描述 Adapter 发现出的一个采集单元（用于生成首页任务）。
type Unit struct {
	Entity   string         `json:"entity"`
	UnitID   string         `json:"unit_id"`
	UnitName string         `json:"unit_name"`
	Params   map[string]any `json:"params,omitempty"`
}

// Home 由采集单元生成首页任务。
func Home(u Unit, source string, page int, targetPages int, jobID string) *Payload {
	if page < 1 {
		page = 1
	}
	p := &Payload{
		JobID:     jobID,
		Source:    source,
		Entity:    u.Entity,
		Operation: OpList,
		Cursor: Cursor{
			UnitID:   u.UnitID,
			UnitName: u.UnitName,
			Page:     page,
		},
	}
	if len(u.Params) > 0 {
		p.Params = u.Params
	}
	if targetPages > 0 {
		p.TargetPages = targetPages
	}
	return p
}

// Next 生成下一页任务（复用同 unit 的源/实体/参数，仅推进 page）。
func (p *Payload) Next() *Payload {
	next := *p
	next.Cursor.Page = p.Cursor.Page + 1
	if p.Params != nil {
		cp := make(map[string]any, len(p.Params))
		for k, v := range p.Params {
			cp[k] = v
		}
		next.Params = cp
	}
	return &next
}

// Valid 校验是否为一个合法的 list 任务。
func (p *Payload) Valid() bool {
	return p != nil &&
		p.Operation == OpList &&
		p.Source != "" &&
		p.Entity != "" &&
		p.Cursor.UnitID != "" &&
		p.Cursor.Page > 0
}

// CursorKey 游标 Hash 键：cur:{source}:{entity}:{unit_id}。
func (p *Payload) CursorKey() string {
	return fmt.Sprintf("cur:%s:%s:%s", p.Source, p.Entity, p.Cursor.UnitID)
}

// NextPageLockKey 幂等“追加下一页”闸门键：addlock:{source}:{entity}:{unit_id}:{page}。
func (p *Payload) NextPageLockKey() string {
	return fmt.Sprintf("addlock:%s:%s:%s:%d", p.Source, p.Entity, p.Cursor.UnitID, p.Cursor.Page)
}

// Describe 可读日志描述：source[unit_name] p{page}。
func (p *Payload) Describe() string {
	unit := p.Cursor.UnitName
	if unit == "" {
		unit = p.Cursor.UnitID
	}
	return fmt.Sprintf("[%s|%s|p%d]", p.Source, unit, p.Cursor.Page)
}

// TargetPages 目标页上限：显式 TargetPages 优先，否则用 fallback。
func (p *Payload) TargetPagesOr(fallback int) int {
	if p.TargetPages > 0 {
		return p.TargetPages
	}
	return fallback
}

// Encode 序列化为 Redis 消息体。
func (p *Payload) Encode() ([]byte, error) {
	return json.Marshal(p)
}

// Decode 从 Redis 消息体解析任务；兼容 JSON 字符串二次包装的情况。
func Decode(raw []byte) (*Payload, error) {
	var p Payload
	if err := json.Unmarshal(raw, &p); err != nil {
		// 站点/中间件偶发把 JSON 再包一层字符串
		var s string
		if err2 := json.Unmarshal(raw, &s); err2 == nil {
			if err3 := json.Unmarshal([]byte(s), &p); err3 == nil {
				return &p, nil
			}
		}
		return nil, err
	}
	return &p, nil
}
